robots/flakefinder/downloader.go

```diff
@@ -0,0 +1,197 @@
+	dirOfJobs := path.Join(dirOfPrJobs, job)
+
+	prJobs, err := listGcsObjects(ctx, client, bucket, dirOfJobs+"/", "/")
+	if err != nil {
+		return nil, fmt.Errorf("error listing gcs objects: %v", err)
+	}
+	builds := sortBuilds(prJobs)
+	profilePath := ""
+	buildNumber := 0
+	reports := []*Result{}
+	for _, build := range builds {
+		buildDirPath := path.Join(dirOfJobs, strconv.Itoa(build))
+		dirOfFinishedJSON := path.Join(buildDirPath, finishedJSON)
+		dirOfStartedJSON := path.Join(buildDirPath, startedJSON)
+
+		_, err := readGcsObject(ctx, client, bucket, dirOfFinishedJSON)
+		if err == storage.ErrObjectNotExist {
+			// build still running?
+			continue
+		} else if err != nil {
+			return nil, fmt.Errorf("Cannot read finished.json (%s) in bucket '%s'", dirOfFinishedJSON, bucket)
+		} else {
+			startedJSON, err := readGcsObject(ctx, client, bucket, dirOfStartedJSON)
+			if err != nil {
+				return nil, fmt.Errorf("Cannot read started.json (%s) in bucket '%s'", dirOfStartedJSON, bucket)
+			}
+			if !isLatestCommit(startedJSON, pr) {
+				break
+			}
+			buildNumber = build
+			artifactsDirPath := path.Join(buildDirPath, "artifacts")
+			profilePath = path.Join(artifactsDirPath, "junit.functest.xml")
+			data, err := readGcsObject(ctx, client, bucket, profilePath)
+			if err == storage.ErrObjectNotExist {
+				return nil, nil
+			}
+			if err != nil {
+				return nil, err
+			}
+			report, err := junit.Ingest(data)
+			if err != nil {
+				return nil, err
+			}
+			reports = append(reports, &Result{Job: job, JUnit: report, BuildNumber: buildNumber, PR: pr.Number})
+		}
+	}
+
+	return reports, nil
+}
+
+// sortBuilds converts all build from str to int and sorts all builds in descending order and
+// returns the sorted slice
+func sortBuilds(strBuilds []string) []int {
+	var res []int
+	for _, buildStr := range strBuilds {
+		num, err := strconv.Atoi(buildStr)
+		if err != nil {
+			logrus.Infof("Non-int build number found: '%s'", buildStr)
+		} else {
+			res = append(res, num)
+		}
+	}
+	sort.Sort(sort.Reverse(sort.IntSlice(res)))
+	return res
+}
+
+type finishedStatus struct {
+	Timestamp int
+	Passed    bool
+}
+
+// {"timestamp":1562772668,"pull":"2473","repo-version":"f3bb83f4377b8b45bd47d33373edfacf85361f0e","repos":{"kubevirt/kubevirt":"release-0.13:577e95c340e1b21ff431cbba25ad33c891554e38,2473:8c33c116def661c69b4a8eb08fac9ca07dfbf03c"}}
+type startedStatus struct {
+	Timestamp int
+	Repos     map[string]string
+}
+
+type Result struct {
+	Job         string
+	JUnit       []junit.Suite
+	BuildNumber int
+	PR int
+}
+
+func isLatestCommit(jsonText []byte, pr *github.PullRequest) bool {
+	var status startedStatus
+	err := json.Unmarshal(jsonText, &status)
+	for _, v := range status.Repos {
+		return err == nil && strings.Contains(v, fmt.Sprintf("%d:%s", pr.Number, pr.Head.SHA))
```

Comment:
A nit here:

IIUC you're leveraging that if ranging over an empty slice the `for` block will not get executed. if it is, the `err` and first element are checked. I find this puzzling and harder to read.

Is this a normal idiom of go? If not I'd rather use `if len(status.Repos) == 0` to guard against an empty slice here. Then afterwards I'd check the first element. This would make it more obvious what exactly is happening here.

Reply:
I should have iterated over all values and not just the first one, so this is a mistake on my side (would only have noticed it when working on jobs from the merge pool). In general iterating over `nil` arrays is normal in go.